google/cloud/firestore_v1/stream_generator.py

```diff
@@ -38,3 +65,35 @@ def throw(self, exp=None):
 
     def close(self):
         return self._generator.close()
+    
+    @property
+    def explain_options(self) -> ExplainOptions:
+        """Query profiling options for this stream request."""
+        return self._explain_options
+
+    @property
+    def explain_metrics(self) -> ExplainMetrics:
+        """
+        Get the metrics associated with the query execution.
+        Metrics are only available when explain_options is set on the query. If
+        ExplainOptions.analyze is False, only plan_summary is available. If it is
+        True, execution_stats is also available.
+        :rtype: :class:`~google.cloud.firestore_v1.query_profile.ExplainMetrics`
+        :returns: The metrics associated with the query execution.
+        :raises: :class:`~google.cloud.firestore_v1.query_profile.QueryExplainError`
+            if explain_metrics is not available on the query.
+        """
+        if self._explain_metrics is not None:
+            return self._explain_metrics
+        elif self._explain_options is None:
+            raise QueryExplainError("explain_options not set on query.")
+        elif self._explain_options.analyze is False:
+            # we need to run the query to get the explain_metrics
+            try:
+                next(self)
```

Comment:
This structure is going to cause issues when you make the async version. `@property` methods can't call coroutines.

Maybe we should just raise an exception if this is accessed before it's iterated? Or do you have another idea?

Reply:
> Maybe we should just raise an exception if this is accessed before it's iterated? Or do you have another idea?

With `explain_options.analyze == False`, the query isn't really run, and there's no query result returned. Thus users won't organically iterate the results themselves. I'm not the most knowledgeable person about async python, but it seems possible to make an async property, but it's not recommended isn't very convenient to use. (see [stackoverflow thread](https://stackoverflow.com/questions/54984337/how-should-you-create-properties-when-using-asyncio).) 

Apart from forcing an async property, other options I can think of:
1. Move the logic to inside `Query._make_stream()` ([here](https://github.com/googleapis/python-firestore/blob/a0d2d7e26bd2008cc1f19727002a8c663bce78ec/google/cloud/firestore_v1/query.py#L362)): if `analyze == False`, get `explain_metrics` right after iterator is created, so StreamGenerator doesn't need to call the API anymore. This may make creating the generator a bit slower, though.
2. make explain_metrics not a property but a function. At least formally it's not a property anymore - but this will create a difference between datastore and firestore SDKs

I feel like (1) is a better idea. WDYT?

Reply:
I don't think 1 solves the problem. In that case, _make_stream will be an AsyncGenerator, and you'll need to await generator.__anext__ to get the explain_metrics value from it. (Or maybe I'm misunderstanding the details, but fundamentally, you'll have to await something somewhere to get that value, so we can't use a sync property)

I think 2 is cleaner. Turn it into `get_explain_metrics()`. I think it's ok that it's different from datastore here

Reply:
For 1, I meant we get `explain_metrics` when  `Query._make_stream()` is called, and pass the value to the init function of StreamGenerator, so `explain_metrics` will be strictly a property without any operation to get data when called

Reply:
I don't think we can, because `self._make_stream` is just making an AsyncGenerator when it is [called for the first time](https://github.com/googleapis/python-firestore/blob/cb693f8b02434de803ce41e72bb5aa18c5de3764/google/cloud/firestore_v1/async_query.py#L401C32-L401C44). Basically like an __init__ function. It's code isn't executed until __anext__ is called on it.

If AsyncQuery.stream were an async function (aka coroutine), then we could do something there. But as-is, we're not in an async context at any point there

Reply:
Wow I see, didn't know that `AsyncQuery.stream()` isn't async. Then I will go with option 2.

Reply:
done